### PR TITLE
Tailwind Setup: Use fully specified paths

### LINKED ISSUE #1643: [cli] setup tailwind path issue
This line throws an error when running `yarn rw setup tailwind` in any directory other than root. 

https://github.com/redwoodjs/redwood/blob/db6593e479be88bf4264ef976ce24277e4068ddb/packages/cli/src/commands/setup/tailwind/tailwind.js#L160

I would have assumed that the CLI should be able to handle path errors gracefully.

Maybe this command was just missed?

### PR DESCRIPTION
**Pull request by Tobbe**
Fixes #1643

### CODE REVIEW

**Comment by Tobbe** [comment]
```terminal
tobbe@XPS9550 MINGW64 ~/dev/redwood/tailwindpathspr/web/src/pages/TestPage
$ yarn rw setup tailwind
yarn run v1.22.4
$ C:\Users\tobbe\dev\redwood\tailwindpathspr\node_modules\.bin\rw setup tailwind
  √ Installing packages...
  √ Configuring PostCSS...
  √ Initializing Tailwind CSS...
  √ Adding imports to index.css...
  √ One more thing...

           Tailwind configured with "upcoming change" opt-in enabled

           See this doc for info: https://tailwindcss.com/docs/upcoming-changes

Done in 95.67s.
```

**Comment by Tobbe** [comment]
Fixing this uncovered another bug with the tailwind setup. `--force` isn't working as it should.

```
tobbe@XPS9550 MINGW64 ~/dev/redwood/tailwindpaths/web/src/pages/TestPage
$ yarn rw setup tailwind --force
yarn run v1.22.4
$ C:\Users\tobbe\dev\redwood\tailwindpaths\node_modules\.bin\rw setup tailwind --force
  √ Installing packages...
  √ Configuring PostCSS...
  × Initializing Tailwind CSS...
    → info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
    Adding imports to index.css...
    One more thing...
Command failed with exit code 1: yarn tailwindcss init
   � tailwind.config.js already exists.
error Command failed with exit code 1.
$ C:\Users\tobbe\dev\redwood\tailwindpaths\node_modules\.bin\tailwindcss init

   @tailwindcss/postcss7-compat 2.0.2


info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Done in 99.95s.
```

I'll try to fix that once this PR is merged

**Comment by thedavidprice** [comment]
> Fixing this uncovered another bug with the tailwind setup. --force isn't working as it should.

@Tobbe please do take a look at the original PR addressing `--force` with this command. I seem to recall needing to fail the overwrite if/when there was a case where manual changes to the tailwind config existed. Maybe? Anyway, it might not be more output messaging confusion vs. command not working. Will defer to you.

**Review by thedavidprice — APPROVED** [review]
LGTM! Let me know if you need a Mac use to test first before merging. Otherwise 🚀

**Comment by Tobbe** [comment]
> Let me know if you need a Mac use to test first before merging.

That would be great! This is how I tested it:

```
yarn create redwood-app twpath
cd twpath
yarn rw upgrade --pr 1648:0.23.0-8487f8c
yarn rw g page test
cd web/src/pages/TestPage
yarn rw setup tailwind
```

**Comment by Tobbe** [comment]
I went back and took another look at this. What a mess the tailwind installation is!
For me I actually have two installations. One in ~/.../twpaths/node_modules and another one in ~/.../twpaths/web/node_modules

So if I'm running the setup command in ~/.../twpaths it will use the installation in ~/.../twpaths/node_modules, and files will be created relative to that path. If I'm somewhere inside web/, like ~/.../twpaths/web/src/pages, it will use the installation in ~/.../twpaths/web/node_modules and files will be created relative to that path instead.

**Comment by Tobbe** [comment]
> please do take a look at the original PR addressing `--force`

You mean this one? https://github.com/redwoodjs/redwood/issues/1301 I don't see anything in there about blocking `--force`. What am I missing?

# Here's my understanding of the issue I found:

**Scenario 1 (works)**
```
~/.../twpath $ yarn rw setup tailwind
```
`yarn tailwindcss init` is executed. It creates `~/.../twpath/tailwind.config.js`

At this point, if you were to run `yarn tailwindcss init` again it would fail, complaining about the existence of `tailwind.config.js`

Our setup command moves the config file to `web/tailwind.config.js`

```
~/.../twpath $ yarn rw setup tailwind --force
```
`web/tailwind.config.js` exists, but because `--force` is specified we continue
`yarn tailwindcss init` is executed. It doesn't find any config file (because we previously moved it into `web/`), and so it happily creates a new one in `~/.../twpath`. We will move it into `web/`, overwriting the existing file

**Scenario 2 (fails)**
```
~/.../twpath/web/src/pages/TestPage $ yarn rw setup tailwind
```
`yarn tailwindcss init` is executed. `cwd()` is now `~/.../twpath/web`, so it creates `~/.../twpath/web/tailwind.config.js`

At this point, if you were to run `yarn tailwindcss init` again it would fail, complaining about the existence of `tailwind.config.js`

Our setup command moves the config file to `web/tailwind.config.js` (which is a no-op, since that's where it already is)

```
~/.../twpath/web/src/pages/TestPage $ yarn rw setup tailwind --force
```
`web/tailwind.config.js` exists, but because `--force` is specified we continue
`yarn tailwindcss init` is executed. It now finds `web/tailwind.config.js` because `cwd()` is `~/.../twpath/web`. The command fails, complaining about the existing config file.

File: packages/cli/src/commands/setup/tailwind/tailwind.js
Comment (peterp) [inline packages/cli/src/commands/setup/tailwind/tailwind.js:151]: I think it should definitely live in "web"

File: packages/cli/src/commands/setup/tailwind/tailwind.js
Comment (peterp) [inline packages/cli/src/commands/setup/tailwind/tailwind.js:165]: I actually don't think this comment is true anymore. I think it must live in the root of "web" in order to be picked up by the extension.

File: packages/cli/src/commands/setup/tailwind/tailwind.js
Comment (Tobbe) [reply packages/cli/src/commands/setup/tailwind/tailwind.js:151]: Yes, that's why we move it later on, into web/ (if it isn't there already)

File: packages/cli/src/commands/setup/tailwind/tailwind.js
Comment (Tobbe) [reply packages/cli/src/commands/setup/tailwind/tailwind.js:165]: Exactly, what's why we have it there for the time being. In the future, if it becomes possible to tell the extension where to look, we can move it into web/config/

File: packages/cli/src/commands/setup/tailwind/tailwind.js
Comment (Tobbe) [reply packages/cli/src/commands/setup/tailwind/tailwind.js:151]: This comment might give you more context https://github.com/redwoodjs/redwood/pull/1648#issuecomment-761805959

File: packages/cli/src/commands/setup/tailwind/tailwind.js
Comment (peterp) [reply packages/cli/src/commands/setup/tailwind/tailwind.js:165]: Aaaaah-ha!

**Review by peterp — APPROVED** [review]

File: packages/cli/src/commands/setup/tailwind/tailwind.js
Comment (peterp) [reply packages/cli/src/commands/setup/tailwind/tailwind.js:151]: I see!

Comment (peterp) [reply packages/cli/src/commands/setup/tailwind/tailwind.js:151]: I think the path where it says "will handle this" wasn't very to clear to me. Maybe we could say something like: If the user does not specify the path the config will be moved to `./web/tailwind.config.js`

Comment (Tobbe) [reply packages/cli/src/commands/setup/tailwind/tailwind.js]: The user can't specify the path. I tried rewriting the comment. Is it clearer now?

Comment (peterp) [reply packages/cli/src/commands/setup/tailwind/tailwind.js]: Yeah! I assumed that fs was using `cwd`, but it was a few mental jumps to make that assumption, maybe you could make this explicit by using `path.join(process.cwd(), 'tailwind.config.js')` instead, which makes the code is easier to read and then the comment isn't required!

Comment (Tobbe) [reply packages/cli/src/commands/setup/tailwind/tailwind.js]: Hahah. Like this? https://github.com/redwoodjs/redwood/commit/b0174ef840b9250b9f72b22a59d522934eee9409 😆

I had that, but then removed it to "simplify" the code. But I absolutely agree that expressive code is better than comments. I'll change it back again

**Comment by Tobbe** [comment]
@thedavidprice If you could run a quick test on your (mac) computer, that would be awesome!

```
yarn create redwood-app twpath
cd twpath
yarn rw upgrade --pr 1648:0.23.0-e3c2e45
yarn rw g page test
cd web/src/pages/TestPage
yarn rw setup tailwind
```

**Review by thedavidprice — CHANGES_REQUESTED** [review]
@Tobbe this is working (almost) 💯 on Mac! From root, install and --force work correctly. From a sub-directory, install works correctly but --force runs into an issue with Tailwind `yarn tailwindcss init` that fails if `tailwind.config.js` exists. This is the reason why we added the "Sync ..." step that runs `yarn install --check-files`, which avoids Tailwind throwing. Here's the error:
```console
Command failed with exit code 1: yarn tailwindcss init
   🚫 tailwind.config.js already exists.
error Command failed with exit code 1.
$ /Users/price/Repos/xx-delete/node_modules/.bin/tailwindcss init

   @tailwindcss/postcss7-compat 2.0.2
```

To run --force successfully from any directory, I think we need to set the Execa working directory to root on this line:
https://github.com/redwoodjs/redwood/blob/ab1b7229f40dd6668244f1bb2d866ed8829de806/packages/cli/src/commands/setup/tailwind/tailwind.js#L89

**Comment by peterp** [comment]
@Tobbe Thinking out loud here, maybe we can pass `cwd` to execa and always make it use `getPaths().web.base`, then we can remove the "move" code, since we would always init the tailwind in the right place.

**Comment by Tobbe** [comment]
> @Tobbe this is working (almost) 💯 on Mac!

Thanks for the thorough testing. You're seeing the exact same thing as I'm seeing on Windows. You can compare your output with what I posted in a previou comment https://github.com/redwoodjs/redwood/pull/1648#issuecomment-760747911

As I also said in that comment, my plan was to merge this fix, and then write a separate fix for the --force stuff. But I guess I can do it all here.

My plan for fixing --force was to just manually delete `tailwind.config.js` first, before running `yarn tailwindcss init`.

> To run --force successfully from any directory, I think we need to set the Execa working directory to root on this line:
> ` await execa('yarn', ['install', '--check-files']) `
@thedavidprice We get "tailwind.config.js already exists." from `tailwindcss init`. Will `yarn install --check-files` help with that?

**Comment by thedavidprice** [comment]
Ah, roger that @Tobbe I (falsely) assumed you were ready to go and skipped over:
> You're seeing the exact same thing as I'm seeing on Windows. You can compare your output with what I posted in a previous comment #1648 (comment)

Yes, `yarn install --check-files` handles the "previously installed" error. This is why `--force` works when run from root. So my vote would be to go with @peterp suggestion to set `cwd`.

Thanks! 'Cause this week you are:
on-fire

**Comment by Tobbe** [comment]
> This is why `--force` works when run from root.

I don't think that's true. The reason it works from root is that there is no tailwind.config.js file there that `yarn tailwindcss init` can find. You can try it yourself by manually running `yarn tailwindcss init` in root. It will work just fine. Then go in to web/ and run it from there, and it will give you that same error message as you get when running `yarn rw setup tailwind --force`

**Comment by Tobbe** [comment]
No luck with that

```
tobbe@XPS9550 MINGW64 ~/dev/redwood/tailwindpathspr
$ yarn rw setup tailwind --force
yarn run v1.22.4
$ C:\Users\tobbe\dev\redwood\tailwindpathspr\node_modules\.bin\rw setup tailwind --force
  √ Installing packages...
  √ Configuring PostCSS...
  × Initializing Tailwind CSS...
    → spawn C:\WINDOWS\system32\cmd.exe ENOENT
    Adding imports to index.css...
    One more thing...
Command failed with ENOENT: yarn tailwindcss init
spawn C:\WINDOWS\system32\cmd.exe ENOENT
Done in 101.18s.
```

**Comment by Tobbe** [comment]
@thedavidprice As I expected, now, when I specify cwd to be web/ it doesn't even work with --force in root anymore, because now `yarn tailwindcss init` finds tailwind.config.js in web/, even if we stand in root.

```
tobbe@XPS9550 MINGW64 ~/dev/redwood/tailwindpathspr
$ yarn rw setup tailwind --force
yarn run v1.22.4
$ C:\Users\tobbe\dev\redwood\tailwindpathspr\node_modules\.bin\rw setup tailwind --force
  √ Installing packages...
  √ Configuring PostCSS...
  × Initializing Tailwind CSS...
    → info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
    Adding imports to index.css...
    One more thing...
Command failed with exit code 1: yarn tailwindcss init
   � tailwind.config.js already exists.
error Command failed with exit code 1.
$ C:\Users\tobbe\dev\redwood\tailwindpathspr\node_modules\.bin\tailwindcss init

   @tailwindcss/postcss7-compat 2.0.2


info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Done in 115.82s.
```

**Comment by Tobbe** [comment]
Now that I delete the config file first it works

```
tobbe@XPS9550 MINGW64 ~/dev/redwood/tailwindpathspr
$ yarn rw setup tailwind --force
yarn run v1.22.4
$ C:\Users\tobbe\dev\redwood\tailwindpathspr\node_modules\.bin\rw setup tailwind --force
  √ Installing packages...
  √ Configuring PostCSS...
  √ Initializing Tailwind CSS...
  ↓ Adding imports to index.css... [skipped]
    → Imports already exist in index.css
  √ One more thing...

           Tailwind configured with "upcoming change" opt-in enabled

           See this doc for info: https://tailwindcss.com/docs/upcoming-changes

Done in 113.02s.
```

**Comment by Tobbe** [comment]
Next problem to handle: Paths with spaces!

```
tobbe@XPS9550 MINGW64 ~/dev/redwood/tail wind pathspr
$ yarn rw setup tailwind --force
yarn run v1.22.4
$ "C:\Users\tobbe\dev\redwood\tail wind pathspr\node_modules\.bin\rw" setup tailwind --force
  √ Installing packages...
  √ Configuring PostCSS...
  × Initializing Tailwind CSS...
    → spawn C:\WINDOWS\system32\cmd.exe ENOENT
    Adding imports to index.css...
    One more thing...
Command failed with ENOENT: yarn tailwindcss init
spawn C:\WINDOWS\system32\cmd.exe ENOENT
Done in 102.09s.
```

**Review by peterp — APPROVED** [review]

**Comment by Tobbe** [comment]
🎉 

```
tobbe@XPS9550 MINGW64 ~/dev/redwood/tail wind pathspr
$ yarn rw setup tailwind --force
yarn run v1.22.4
$ "C:\Users\tobbe\dev\redwood\tail wind pathspr\node_modules\.bin\rw" setup tailwind --force
  √ Installing packages...
  √ Configuring PostCSS...
  √ Initializing Tailwind CSS...
  ↓ Adding imports to index.css... [skipped]
    → Imports already exist in index.css
  √ One more thing...

           Tailwind configured with "upcoming change" opt-in enabled

           See this doc for info: https://tailwindcss.com/docs/upcoming-changes

Done in 128.52s.

tobbe@XPS9550 MINGW64 ~/dev/redwood/tail wind pathspr
$ cd web/src/pages/

tobbe@XPS9550 MINGW64 ~/dev/redwood/tail wind pathspr/web/src/pages
$ yarn rw setup tailwind --force
yarn run v1.22.4
$ "C:\Users\tobbe\dev\redwood\tail wind pathspr\node_modules\.bin\rw" setup tailwind --force
  √ Installing packages...
  √ Configuring PostCSS...
  √ Initializing Tailwind CSS...
  ↓ Adding imports to index.css... [skipped]
    → Imports already exist in index.css
  √ One more thing...

           Tailwind configured with "upcoming change" opt-in enabled

           See this doc for info: https://tailwindcss.com/docs/upcoming-changes

Done in 118.56s.
```

Works with spaces in the path, and both at root level and deep inside web/

**Comment by thedavidprice** [comment]
> As I expected, now, when I specify cwd to be web/ it doesn't even work with --force in root anymore...

Ah, understood. I was suggesting to set the cwd to be root. All good.

There were two reasons (if I recall correctly) why we had to run the `--check-file`, which were both related to tailwind already being installed. I'm going to re-test now --> there's a chance with you new code we don't need the "Sync" step and can remove the `--check-file` process, which takes way too long anyway.

Testing now.

**Comment by thedavidprice** [comment]
Testing looks good ✅

@Tobbe We don't need this step anymore. Can you remove starting L82?
```
{
            title: 'Sync yarn.lock and node_modules',
            task: async () => {
              /**
               * Sync yarn.lock file and node_modules folder.
               * Refer https://github.com/redwoodjs/redwood/issues/1301 for more details.
               */
              await execa('yarn', ['install', '--check-files'])
            },
          },
```

**Comment by Tobbe** [comment]
@thedavidprice I actually already tried that, but just by manually editing the setup command inside node_modules. This is what I got

```
$ yarn rw setup tailwind --force
yarn run v1.22.4
$ "C:\Users\tobbe\dev\redwood\tail wind pathspr\node_modules\.bin\rw" setup tailwind --force
  √ Installing packages...
  √ Configuring PostCSS...
  × Initializing Tailwind CSS...
    → info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
    Adding imports to index.css...
    One more thing...
Command failed with exit code 1: yarn tailwindcss init
error Command "tailwindcss" not found.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Done in 29.75s.
```

But since you say it'll work without it I'm going to build a proper new package to test with

**Comment by Tobbe** [comment]
Same error with the PR package

```
tobbe@XPS9550 MINGW64 ~/dev/redwood/tail wind pathspr
$ yarn rw setup tailwind --force
yarn run v1.22.4
$ "C:\Users\tobbe\dev\redwood\tail wind pathspr\node_modules\.bin\rw" setup tailwind --force
  √ Installing packages...
  √ Configuring PostCSS...
  × Initializing Tailwind CSS...
    → info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
    Adding imports to index.css...
    One more thing...
Command failed with exit code 1: yarn tailwindcss init
error Command "tailwindcss" not found.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Done in 25.39s.
```

**Comment by thedavidprice** [comment]
Super weird. I tested first locally by removing it from my installed package. 

Well then, let's ship this as you had it! Thanks for trying it out. 

And, oh my, it's unreal how much time and attention this specific Setup command has required. This is either round 3 or 4 of a deep dive. We must really ❤️ Tailwind...

**Comment by Tobbe** [comment]
> We must really ❤️ Tailwind...

Haha. I don't even use tailwind. I just wanted to do the community a service 🤷‍♂️ 


I'm running the latest PR package again, on a fresh install, just to rule out my manual changes as the source of the error. Will report back in a bit.

**Comment by Tobbe** [comment]
First run without `yarn install --check-files` works fine. Second one (with --force) throws the error

```
tobbe@XPS9550 MINGW64 ~/dev/redwood/twpath
$ yarn rw setup tailwind
yarn run v1.22.4
$ C:\Users\tobbe\dev\redwood\twpath\node_modules\.bin\rw setup tailwind
  √ Installing packages...
  √ Configuring PostCSS...
  √ Initializing Tailwind CSS...
  √ Adding imports to index.css...
  √ One more thing...

           Tailwind configured with "upcoming change" opt-in enabled

           See this doc for info: https://tailwindcss.com/docs/upcoming-changes

Done in 27.78s.

tobbe@XPS9550 MINGW64 ~/dev/redwood/twpath
$ yarn rw setup tailwind --force
yarn run v1.22.4
$ C:\Users\tobbe\dev\redwood\twpath\node_modules\.bin\rw setup tailwind --force
  √ Installing packages...
  √ Configuring PostCSS...
  × Initializing Tailwind CSS...
    → info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
    Adding imports to index.css...
    One more thing...
Command failed with exit code 1: yarn tailwindcss init
error Command "tailwindcss" not found.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Done in 21.98s.
```

**Comment by thedavidprice** [comment]
🚀

**Pull request merged**
